### DISCUSSION TITLE
Temperature System Improvement

### DISCUSSION THREAD

**Discussion by Mazin-O3** [lcgamboa/picsimlab#93] [General]
Hello, I would like to suggest adding the following features to the Temperature System:
- Temperature range selection like 0-6 °C
- Temperature can be set to rise at a desired rate
- Variable heater and cooler control
- Temperature keeps rising or falling to when heater or cooler is activated

**Comment by lcgamboa** [comment]
Hi @MazinCE ,

Currently, temperature control is implemented with the[ first-order system discrete transfer function](https://github.com/lcgamboa/picsimlab/blob/5ed5ad7b58f552c83a1a5f047f215a9741d3657f/src/parts/other_TempSys.cc#L188). This transfer function was obtained through tests on real hardware. The calculation takes into account the ambient temperature of 27.5°C. Is this value you suggest changing from 0 to 6°C?
To control the temperature variation rate it would be interesting to model in relation to the value of the resistor used and the fan flow used instead of simply changing the coefficients. (There is already a spare part [D. Transfer Function](https://lcgamboa.github.io/picsimlab_docs/stable/DTransferFunction.html) that can be used to model any system)
Heating and ventilation control is already variable, just apply a PWM signal to the inputs.
As it is modeled as a first-order system, the temperature will reach a maximum value. As for the minimum value, it makes no sense for the temperature to be lower than the ambient temperature using only one fan.

**Reply by Mazin-O3** [reply]
Ahh I see, 
Yes I think it would be nice to have a configurable temperature range instead of the fixed 27.5°C. I have tried applying PWM but I couldn't quite get it to reach the maximum allowed temperature gradually. I'm trying to use it as thermostat/compressor for my fridge project.

**Comment by lcgamboa** [comment]
I added the option to change the room temperature to the task list.
Maybe create a new spare part with a Thermoelectric
Cooler (type TEC1-12706) and a temperature sensor is a more suitable option for your case.
Have you done any mathematical modeling of your fridge's thermostat/compressor system?

**Reply by Mazin-O3** [reply]
Thank you! That would be great. I haven't delved into mathematical modeling as I prefer to keep things simple at the moment.

**Comment by lcgamboa** [comment]
Without the mathematical model, it's not possible to simulate.  The modeling it's not difficult, take a look at [this article](https://www.laccei.org/LACCEI2020-VirtualEdition/full_papers/FP51.pdf) for example.

**Reply by Mazin-O3** [reply]
Thank you, I really appreciate it.

**Comment by lcgamboa** [comment]
I have added support to change the ambient temperature in the Temperature system spare part in the[ last build ](https://github.com/lcgamboa/picsimlab/releases/tag/latestbuild)version.
